### PR TITLE
Add CredentialsProvider interface to allow the use of time based credentials

### PR DESCRIPTION
**Pull request by kampka**
It's become increasingly popular for infrastructure providers and security engineers to push for the use of time base credentials for connecting to the database instead of having long lasting secrets hard coded in the database.
 Examples include, among others, [AWS](https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/UsingWithRDS.IAMDBAuth.html) or [HashiCorp VAULT](https://developer.hashicorp.com/vault/docs/secrets/databases/postgresql).

The connection flow is usually like this: When a database driver needs to establish a connection, it will fetch fresh credentials on demand. These credentials are valid for a short time frame, usually minutes, which is long enough for the driver to establish a connection. If the driver ever needs to re-connect, it needs to obtain a fresh set of credentials.
Since most database drivers provide thread pools for handling bare connections, the challenge now lies with making the thread pool aware of the fact that fresh credentials are required and how to obtain them.

This PR extends the `ConnectionPoolConfiguration` with the possibility to provide a `CredentialsProvider` that can be implemented by the user to pass in a custom source for credentials. The `ConnectionFactory` then resolves a fresh configuration every time it establishes a new connection.
I tried to extend all interfaces in a way that should preserve compatibility in nearly all cases.

The `ktLint` task seems to be broken or not setup properly, and running `ktlint -F` on the code base produces a lot of changes all over. I therefore decided to forgo automatic formatting. Hope this is OK.

### CODE REVIEW

**Comment by oshai** [comment]
Thanks for the PR. It will take me a couple of weeks to get to review it at the moment.

**Comment by kampka** [comment]
Sure thing. Let me know if anything needs clarification.

**Comment by kampka** [comment]
> Can you provide some more details on drivers implementing that or such practices?

It's not usually implemented in the driver but in the connection pool, which is often a separate thing, eg. when using JDBC.
An example of how to implement it in Hikari can be found here: https://blog.jdriven.com/2021/06/configure-hikari-connection-pool-when-using-rds-iam/

This example has two "advantages" when dealing with this problem. Being in the JDBC driver space, they don't care to much about non-blocking programming, so they can leverage the existing `getPassword` method to retrieve a token by doing a blocking network request against the AWS API to retrieve the password.
I would argue that doing blocking network requests is not desirable for our case, I opted for introducing a new interface via the provider.

> Basically I think that having both user+password and a provider in the pool config is confusing, so we have to figure another way to tackle this (without breaking compatibility).

I completely agree, but I am not sure how that can be achieved, given that the Kotlin devs opted for a function coloring approach when building in the `suspend` support the way that they did.
Given that the password needs to be retrieved in a timely fashion, it's not feasible to do it once and then keep it around in an immutable configuration object.
So the cleanest idea I can currently come up with is to deprecate the username / password options on the `ConnectionPoolConfiguration` in favor of the `CredentialsProvider` interface. That would give folks at least a warning during compile time and the way it is implemented should already fail sufficiently often when using both at the same time due to the lazy initialization check.

**Comment by kampka** [comment]
Looks like running `ktlintFormat` with JDK 17 causes it to fail with a reflection error.
Not sure if that is anything that can be addressed in this project, but at least the formatting should be correct now (using JDK 11).

**Comment by oshai** [comment]
> deprecate the username / password options

I don't think that's a good alternative as this is the way most users use the lib.
Maybe we can have a default provider that just looks on the user/password.

In addition, instead of configuration inheritence of ConfigurationResolver (which breaks compatibility)
have the resolver as a field.

**Comment by kampka** [comment]
> Maybe we can have a default provider that just looks on the user/password.

Definitely, and this is already how it currently works. If the `connectionPoolConfiguration` is not supplied with a `credentialsProvider`, a `StaticCredentialsProvider` with the configured username and password is constructed on the fly. 

> In addition, instead of configuration inheritence of ConfigurationResolver (which breaks compatibility) have the resolver as a field.

Agreed. I have refactored the initial approach a bit. This moves the use of the `CredentialsProvider` into the respective `ConnectionFactory` but allows to not break the API without the need for the additional `ConfigurationResorver` interface.

**Comment by kampka** [comment]
Thanks for clarifying. I think I got all your suggestions covered now.

File: db-async-common/src/main/java/com/github/jasync/sql/db/Configuration.kt [hunk 1]
```diff
@@ -62,6 +61,7 @@ data class Configuration @JvmOverloads constructor(
     val executionContext: Executor = ExecutorServiceUtils.CommonPool,
     val currentSchema: String? = null,
     val socketPath: String? = null,
+    val credentialsProvider: CredentialsProvider = StaticCredentialsProvider(username, password)
```
Comment (oshai) [inline db-async-common/src/main/java/com/github/jasync/sql/db/Configuration.kt:64]: please add credentialsProvider to javadoc

File: db-async-common/src/main/java/com/github/jasync/sql/db/ConnectionPoolConfiguration.kt [hunk 2]
```diff
@@ -76,7 +75,8 @@ data class ConnectionPoolConfiguration @JvmOverloads constructor(
     val interceptors: List<Supplier<QueryInterceptor>> = emptyList(),
     val maxConnectionTtl: Long? = null,
     val currentSchema: String? = null,
-    val socketPath: String? = null
+    val socketPath: String? = null,
+    val credentialsProvider: CredentialsProvider? = null
```
Comment (oshai) [inline db-async-common/src/main/java/com/github/jasync/sql/db/ConnectionPoolConfiguration.kt:79]: please add credentialsProvider to javadoc

File: db-async-common/src/main/java/com/github/jasync/sql/db/StaticCredentialsProvider.kt [hunk 3]
```diff
@@ -0,0 +1,10 @@
+package com.github.jasync.sql.db
+
+import java.util.concurrent.CompletableFuture
+import java.util.concurrent.CompletionStage
+
+class StaticCredentialsProvider(val username: String, val password: String?) : CredentialsProvider {
```
Comment (oshai) [inline db-async-common/src/main/java/com/github/jasync/sql/db/StaticCredentialsProvider.kt:6]: kotlin allows multiple classes in the same file, so this fits CredentialsProvider.kt

File: mysql-async/src/main/java/com/github/jasync/sql/db/mysql/pool/MySQLConnectionFactory.kt [hunk 4]
```diff
@@ -29,7 +26,17 @@ open class MySQLConnectionFactory(val configuration: Configuration) : Connection
      * @return
      */
     override fun create(): CompletableFuture<MySQLConnection> {
-        val connection = MySQLConnection(configuration = configuration)
-        return connection.connect()
+        return configuration.credentialsProvider
```
Comment (oshai) [inline mysql-async/src/main/java/com/github/jasync/sql/db/mysql/pool/MySQLConnectionFactory.kt:29]: I think instead of this change the connection to use credentialsProvider

File: postgresql-async/src/main/java/com/github/jasync/sql/db/postgresql/pool/PostgreSQLConnectionFactory.kt [hunk 5]
```diff
@@ -12,19 +12,22 @@ private val logger = KotlinLogging.logger {}
...
 class PostgreSQLConnectionFactory(val configuration: Configuration) : ConnectionFactory<PostgreSQLConnection>() {
 
-    init {
-        logger.debug { "PostgreSQLConnectionFactory created with configuration ${configuration.toDebugString()}" }
-    }
-
     override fun create(): CompletableFuture<PostgreSQLConnection> {
-        val connection = PostgreSQLConnection(configuration)
-        return connection.connect()
+        return configuration.credentialsProvider
```
Comment (oshai) [inline postgresql-async/src/main/java/com/github/jasync/sql/db/postgresql/pool/PostgreSQLConnectionFactory.kt:20]: same like in mysql

File: mysql-async/src/test/java/com/github/jasync/sql/db/mysql/util/URLParserSpec.kt [hunk 6]
```diff
@@ -35,15 +35,16 @@ class URLParserSpec {
     fun `create a jdbc mysql connection with the available fields`() {
         val connectionUri = "jdbc:mysql://128.167.54.90:9987/my_database?user=john&password=doe"
 
+        val expected = Configuration(
```
[the remark this answers is not in the document]
Comment (kampka) [reply mysql-async/src/test/java/com/github/jasync/sql/db/mysql/util/URLParserSpec.kt:38]: The tests changes because they test equality on the `Confuguration` object level. Parsing the connection string creates a new `Configuration` instance that has a new `StaticCredentialsProvider` instance defaultet, which does not equal the defaulted instance on the expected `Configuration`.

File: mysql-async/src/main/java/com/github/jasync/sql/db/mysql/pool/MySQLConnectionFactory.kt (hunk 4 above)
Comment (kampka) [reply mysql-async/src/main/java/com/github/jasync/sql/db/mysql/pool/MySQLConnectionFactory.kt:29]: This has some implications regarding refactoring, as the connection currently expects the configuration to be fully resolved on creation, eg. it passes the configuration to the `ConnectionHandler` which in turn expects the username to be present for the handshake message. Given the reactive nature of the `CredentialsProvider`, functionality like this would have to be moved from the constructor to the `onConnect` method. It would probably also cause the change of some public method signatures which are however not part of `Connection` / `ConcreteConnection` interface, so might be OK to do it.

This seems like significant refactoring to me, so I'd like to make sure it's what we want before committing time to it :)

File: mysql-async/src/test/java/com/github/jasync/sql/db/mysql/util/URLParserSpec.kt (hunk 6 above)
[the remark this answers is not in the document]
Comment (oshai) [reply mysql-async/src/test/java/com/github/jasync/sql/db/mysql/util/URLParserSpec.kt:38]: So maybe have CredentialsProvider.equals compare user and password instead?

[the remark this answers is not in the document]
Comment (kampka) [reply mysql-async/src/test/java/com/github/jasync/sql/db/mysql/util/URLParserSpec.kt]: I don't like to mess with object equality if it's not really necessary, so I opted for refactoring the tests to only assert the fields that are relevant to the test instead of comparing the full configuration object.
I find is is cleaner in general.
Hope this makes sense to you.

[the remark this answers is not in the document]
Comment (oshai) [reply mysql-async/src/test/java/com/github/jasync/sql/db/mysql/util/URLParserSpec.kt]: it's convenient for many cases to compare configuration data class. so I think it's better to keep that property...

Also I think it should be a simple change: make StaticCredentialsProvider a data class.

[the remark this answers is not in the document]
Comment (kampka) [reply mysql-async/src/test/java/com/github/jasync/sql/db/mysql/util/URLParserSpec.kt]: Making the `StaticCredentialsProvider` would fix this behavior for a very specific variant of `Configuration` objects, but it's not a general solution. Any other `CredentialsProvider` would probably break this transitive contract unless we enforce it on the interface, which I personally would not do.
If we want to preserve this property of the `Configuration`, it would probably be best to implement a custom `equals()` method that only covers the 'static' fields.

[the remark this answers is not in the document]
Comment (oshai) [reply mysql-async/src/test/java/com/github/jasync/sql/db/mysql/util/URLParserSpec.kt]: Practically, I think most of the users will use `StaticCredentialsProvider` so although it's not a general solution I think it's a good compromise for convenience for the majority of usages.

[the remark this answers is not in the document]
Comment (kampka) [reply mysql-async/src/test/java/com/github/jasync/sql/db/mysql/util/URLParserSpec.kt]: I think this might be obsolete.
By defaulting the credentials provider, we break another feature of the `Configuration` data class, which is the copy constructor. For example, `Configuration().copy(password = 'foobar')` will not currently update the static provider.
I switched the implementation around to a lazy defaulting of the provider, which also solve the issues with the tests as the initial default on the `data class` is now `null`

[the remark this answers is not in the document]
Comment (kampka) [reply mysql-async/src/test/java/com/github/jasync/sql/db/mysql/util/URLParserSpec.kt]: @oshai What you think, can we move forward with this approach? :)

[the remark this answers is not in the document]
Comment (oshai) [reply mysql-async/src/test/java/com/github/jasync/sql/db/mysql/util/URLParserSpec.kt]: sorry I missed that. I think it looks good to merge. Will do that soon unless you have anything to add to the PR

[the remark this answers is not in the document]
Comment (kampka) [reply mysql-async/src/test/java/com/github/jasync/sql/db/mysql/util/URLParserSpec.kt]: No worries, we are not in a hurry after all :)

I think this is good to merge. If you can find the time, I'd appreciate it if you could also release a new version.

**Comment by oshai** [comment]
Thanks! released in 2.1.9.
